File: epi_cpp/04_LinkedLists/03_reverse_sublist.cpp
#include <iostream>
#include <memory>

using namespace std;

template <typename T>
struct ListNode{
    T data;
    shared_ptr<ListNode<T>> next;
};

void PrintList(shared_ptr<ListNode<int>> node){
    while(node){
        cout << node->data << " ";
        node = node->next;
    }
    cout << endl;
}

shared_ptr<ListNode<int>> ReverseSublist(shared_ptr<ListNode<int>> L, int start, int finish){
    if(start == finish){
        return L;
    }
    auto dummy_head = make_shared<ListNode<int>>(ListNode<int>{0, L});
    auto sublist_head = dummy_head;
    int k=1;
    while(k++ < start){
        sublist_head = sublist_head->next;
    }
    auto sublist_iter = sublist_head->next;
    while(start++ < finish){
        auto temp = sublist_iter->next;
        sublist_iter->next = temp->next;
        temp->next = sublist_head->next;
        sublist_head->next = temp;
    }
    return dummy_head->next;
}

int main(){
    shared_ptr<ListNode<int>> L1 = make_shared<ListNode<int>>(ListNode<int>{0, nullptr});
    auto temp = L1;
    for(int i=1; i <= 10; ++i){
        temp->next = make_shared<ListNode<int>>(ListNode<int>{i, nullptr});
        temp = temp->next;
    }
    PrintList(L1);
    L1 = ReverseSublist(L1, 3, 8);
    PrintList(L1);
    return 0;
}

File: epi_python/04_LinkedLists/03_reverse_sublist.py
from typing import Optional

class ListNode:
    def __init__(self, data=0, next=None) -> None:
        self.data = data
        self.next = next

def print_list(L: ListNode) -> None:
    while L:
        print(L.data)
        L = L.next

def reverse_sublist(L: ListNode, start: int, finish: int) -> Optional[ListNode]:
    dummy_head = sublist_head = ListNode(0, L)
    for _ in range(1, start):
        sublist_head = sublist_head.next

    sublist_iter = sublist_head.next
    for _ in range(finish - start):
        temp = sublist_iter.next
        sublist_iter.next, temp.next, sublist_head.next = (temp.next,
                                                           sublist_head.next,
                                                           temp)
    return dummy_head.next

if __name__ == '__main__':
    L = L_tail = ListNode(1)
    L_tail.next = ListNode(2)
    L_tail = L_tail.next
    L_tail.next = ListNode(3)
    L_tail = L_tail.next
    L_tail.next = ListNode(4)
    L_tail = L_tail.next
    L_tail.next = ListNode(5)
    L_tail = L_tail.next
    L_tail.next = ListNode(6)
    L_tail = L_tail.next
    L_tail.next = ListNode(7)
    L_tail = L_tail.next
    reverse_sublist(L, 3, 6)
    print_list(L)

